### PR TITLE
Support import/export for doc type history cleanup policy

### PR DESCRIPTION
**Pull request by p-m-j**

### CODE REVIEW

**Comment by p-m-j** [comment]
Reminder check behaviour for import when present in database but missing from xml.
This should end up consistent with other properties handled by import e.g. removed property type in xml.

File: src/Umbraco.Core/Packaging/PackageDataInstallation.cs [hunk 1]
```diff
@@ -692,10 +693,45 @@ private IContentType UpdateContentTypeFromXml(XElement documentType, IContentTyp
...
 
+        private void UpdateHistoryCleanupPolicy(IContentType contentType, XElement element)
+        {
+            contentType.HistoryCleanup ??= new HistoryCleanup();
+
+            if (element == null)
+            {
+                contentType.HistoryCleanup = new HistoryCleanup();
+                return;
+            }
```
Comment (ronaldbarendse) [inline src/Umbraco.Core/Packaging/PackageDataInstallation.cs:709]: This should not alter existing configuration if the element is missing:
```suggestion
            if (element == null)
            {
                // Keep existing configuration (if any)
                return;
            }
            
            contentType.HistoryCleanup ??= new HistoryCleanup();
```

File: src/Umbraco.Core/Packaging/PackageDataInstallation.cs [hunk 2]
```diff
@@ -692,10 +693,45 @@ private IContentType UpdateContentTypeFromXml(XElement documentType, IContentTyp
...
+
+            if (int.TryParse(element.Attribute("keepAllVersionsNewerThanDays")?.Value, out var keepAll))
+            {
+                contentType.HistoryCleanup.KeepAllVersionsNewerThanDays = keepAll;
+            }
+            else
+            {
+                contentType.HistoryCleanup.KeepAllVersionsNewerThanDays = null;
+            }
```
Comment (ronaldbarendse) [inline src/Umbraco.Core/Packaging/PackageDataInstallation.cs:723]: Also keep the existing value when this setting isn't provided/valid:
```suggestion
```

File: src/Umbraco.Core/Packaging/PackageDataInstallation.cs [hunk 3]
```diff
@@ -692,10 +693,45 @@ private IContentType UpdateContentTypeFromXml(XElement documentType, IContentTyp
...
+
+            if (int.TryParse(element.Attribute("keepLatestVersionPerDayForDays")?.Value, out var keepLatest))
+            {
+                contentType.HistoryCleanup.KeepLatestVersionPerDayForDays = keepLatest;
+            }
+            else
+            {
+                contentType.HistoryCleanup.KeepLatestVersionPerDayForDays = null;
+            }
```
Comment (ronaldbarendse) [inline src/Umbraco.Core/Packaging/PackageDataInstallation.cs:732]: Also keep the existing value when this setting isn't provided/valid:
```suggestion
```

File: src/Umbraco.Core/Persistence/Repositories/Implement/ContentTypeRepository.cs [hunk 4]
```diff
@@ -297,6 +298,10 @@ protected override void PersistUpdatedItem(IContentType entity)
 
         private void PersistHistoryCleanup(IContentType entity)
         {
+            // historyCleanup property is not mandatory for api endpoint if it's not present instantiate a default one.
+            // DocumentTypeSave doesn't handle this for us like ContentType constructors do.
+            entity.HistoryCleanup ??= new HistoryCleanup();
```
Comment (ronaldbarendse) [inline src/Umbraco.Core/Persistence/Repositories/Implement/ContentTypeRepository.cs:303]: We should not alter the existing entity, so lets replace this code with the changes I did in the `v8/bugfix/historycleanup-cleanup` branch:
 https://github.com/umbraco/Umbraco-CMS/blob/699403c5a715ed4e90f8d477ac835eda984c0277/src/Umbraco.Core/Persistence/Repositories/Implement/ContentTypeRepository.cs#L298-L310

File: src/Umbraco.Core/Services/Implement/EntityXmlSerializer.cs [hunk 5]
```diff
@@ -447,12 +448,17 @@ public XElement Serialize(IContentType contentType)
...
+       
+
             var xml = new XElement("DocumentType",
                 info,
                 structure,
                 genericProperties,
                 tabs);
 
+            // Never null in v8, ContentTypeCommonRepository.MapHistoryCleanup always returns an instance even if no db record.
+            xml.Add(SerializeCleanupPolicy(contentType.HistoryCleanup));
```
Comment (ronaldbarendse) [inline src/Umbraco.Core/Services/Implement/EntityXmlSerializer.cs:460]: The entity might not come from the repository, so lets be safe and skip non-existent values:
```suggestion
            if (contentType.HistoryCleanup != null)
            {
                xml.Add(SerializeCleanupPolicy(contentType.HistoryCleanup));
            }
```

**Review by ronaldbarendse — CHANGES_REQUESTED** [review]
I've added some comments and if we merge the changes from `v8/bugfix/historycleanup-cleanup` into this PR, can verify both XML import/export and using the services with `entity.HistoryCleanup = null` (e.g. when the configuration isn't part of the Web API request and left empty).

File: src/Umbraco.Core/Packaging/PackageDataInstallation.cs (hunk 2 above)
Comment (p-m-j) [reply src/Umbraco.Core/Packaging/PackageDataInstallation.cs:723]: What if you intentionally want to remove the value?

File: src/Umbraco.Core/Packaging/PackageDataInstallation.cs (hunk 3 above)
Comment (p-m-j) [reply src/Umbraco.Core/Packaging/PackageDataInstallation.cs:732]: What if you intentionally want to remove the value?

File: src/Umbraco.Core/Services/Implement/EntityXmlSerializer.cs (hunk 5 above)
Comment (p-m-j) [reply src/Umbraco.Core/Services/Implement/EntityXmlSerializer.cs]: dcbd51cf

File: src/Umbraco.Core/Packaging/PackageDataInstallation.cs (hunk 1 above)
Comment (p-m-j) [reply src/Umbraco.Core/Packaging/PackageDataInstallation.cs]: dcbd51cf

File: src/Umbraco.Core/Packaging/PackageDataInstallation.cs (hunk 2 above)
Comment (ronaldbarendse) [reply src/Umbraco.Core/Packaging/PackageDataInstallation.cs:722]: You're right, if this value is `null`, the attribute won't be exported and should indeed be set to `null` on import again 👍🏻 We only keep the existing configuration when the `HistoryCleanupPolicy` element isn't present in the XML, so you can import document type from older CMS versions without resetting this value back to the default.

**Review by ronaldbarendse — APPROVED** [review]
I've merged the changes from the `v8/bugfix/historycleanup-cleanup` branch I created, tested the import/export and finally cleaned up the notification/translations a bit:

History cleanup settings

**Pull request merged**